File: src/main/java/net/serebryansky/common/filter/RequestIdFilter.java
package net.serebryansky.common.filter;

import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpHeaders;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;
import reactor.util.context.Context;

import java.util.List;
import java.util.UUID;

import static net.serebryansky.common.util.LoggingUtil.logOnEach;

@Slf4j
public class RequestIdFilter implements WebFilter {
    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        ServerHttpRequest request = exchange.getRequest();
        String requestId = getRequestId(request.getHeaders());
        String invoker = getInvoker(request.getHeaders());
        return chain
                .filter(exchange)
                .doOnEach(logOnEach(r -> log.info("{} {}", request.getMethod(), request.getURI())))
                .contextWrite(Context.of("CONTEXT_KEY", requestId, "INVOKER", invoker));
    }

    private String getInvoker(HttpHeaders headers) {
        List<String> requestIdHeaders = headers.get("X-Source");
        return requestIdHeaders == null || requestIdHeaders.isEmpty()
                ? "UNKNOWN"
                : requestIdHeaders.get(0);
    }

    private String getRequestId(HttpHeaders headers) {
        List<String> requestIdHeaders = headers.get("X-Request-ID");
        return requestIdHeaders == null || requestIdHeaders.isEmpty()
                ? UUID.randomUUID().toString()
                : requestIdHeaders.get(0);
    }
}

File: src/main/kotlin/net/serebryansky/common/filter/RequestIdFilter.kt
package net.serebryansky.common.filter

import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.http.HttpHeaders
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import reactor.util.context.Context
import java.util.*

class RequestIdFilter : WebFilter {
    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val request = exchange.request
        val requestId = getRequestId(request.headers)
        val invoker = getInvoker(request.headers)
        return MDC.putCloseable("MDC_KEY", requestId).use {
            MDC.putCloseable("INVOKER_KEY", invoker).use {
                chain
                    .filter(exchange)
                    .contextWrite(Context.of("CONTEXT_KEY", requestId, "INVOKER", invoker))
            }
        }

    }

    private fun getInvoker(headers: HttpHeaders): String {
        val requestIdHeaders = headers["X-Source"]
        return if (requestIdHeaders == null || requestIdHeaders.isEmpty()) "UNKNOWN" else requestIdHeaders[0]
    }

    private fun getRequestId(headers: HttpHeaders): String {
        val requestIdHeaders = headers["X-Request-ID"]
        return if (requestIdHeaders == null || requestIdHeaders.isEmpty()) UUID.randomUUID()
            .toString() else requestIdHeaders[0]
    }

    companion object {
        private val log = LoggerFactory.getLogger(RequestIdFilter::class.java)
    }
}
